nutrition_labels/grant_data_processing.py

```diff
@@ -0,0 +1,119 @@
+        count = 1
+        merged_grant_chunk = ''
+        merge_grant_nums = list()
+        for grant_chunk in grant_num.split(','):
+            if len(grant_chunk) ==2:
+                grant_chunk = '0' + grant_chunk
+            if count == 2:
+                merge_grant_nums.append(merged_grant_chunk + grant_chunk)
+                merged_grant_chunk = ''
+                count = 1
+            elif count == 1:
+                merged_grant_chunk = grant_chunk
+                count = count + 1
+        merge_grant_nums = [nums for nums in merge_grant_nums if nums != '000000']
+        grant_num = merge_grant_nums
+    else:
+        grant_num = [grant_num]
+    for num in grant_num:
+        epmc_list.append({'pmid':row['pmid'],
+                          'code':row['code'],
+                          'grant_number':num})
+
+epmc_df = pd.DataFrame(epmc_list)
+epmc_df = epmc_df[epmc_df['code'].isin([1,2,3])] # Only selecting useful codes
+
+# getting WT grant number
+def remove_useless_string(string):
+    '''
+    cleans the grant descriptions of artifacts such as <br />
+    :param string: description string
+    :return: clean string
+    '''
+
+    soup = BeautifulSoup(string)
+    string_out = soup.get_text()
+    string_out = string_out.strip('\n')
+    string_out = string_out.strip('\xa0')
+    string_out = re.sub('  ','',string_out)
+    return(string_out)
+
+def only_text(string):
+    '''
+    removes non-alphanumeric characters and spaces to increase matching
+    :param string: description string
+    :return: clean string
+    '''
+    string = re.sub(' |\W','',string)
+    return(string)
+
+
+grant_ref = grant_data[['Grant Programme:Title','Internal ID','Description','Award Date']]
+grant_ref['grant_number'] = grant_data['Internal ID'].apply(lambda x: re.sub('/.*','',x))
+grant_ref = pd.merge(grant_ref,epmc_df, how = 'inner', on = 'grant_number')
+grant_ref['Description'] = grant_ref['Description'].apply(remove_useless_string)
+grant_ref['Description'] = grant_ref['Description'].apply(only_text)
+grant_ref = grant_ref.drop_duplicates(subset=['Description','grant_number','pmid'])
+
+# get remaining entrys with the same pmid grant number but different description
+grant_duplicates = grant_ref[grant_ref.duplicated(subset=['pmid','grant_number'], keep = False)]
+remove_rows = grant_duplicates.index.tolist()
+
+# sort remove list from duplicates (the later grants)
+grant_duplicates = grant_duplicates.sort_values('Award Date').drop_duplicates(subset = ['grant_number','pmid'])
+remove_rows = [i for i in remove_rows if i not in grant_duplicates.index.tolist()]
+grant_ref = grant_ref.drop(remove_rows)
+
+# finding abstract without a matching code
+no_ref = epmc_df[~epmc_df['grant_number'].isin(grant_ref['grant_number'].to_list())]
+no_ref = no_ref[~no_ref['pmid'].isin(grant_ref['pmid'].to_list())]
+missing_abstracts = epmc_tags[epmc_tags['pmid'].isin(no_ref['pmid'].to_list())] #I cant find where this grant is at all
```

Comment:
I think this is interesting, maybe more so the grants that weren't found though, but since you don't use it anywhere else, I'd do some print outs. Also set differences could be more clear here. e.g.
```
print("Grants flagged as relevant in EPMC publications not found in grants data: ")
print(set(epmc_df['grant_number']).difference(set(grant_ref['grant_number'])))
print("PMIDs flagged as relevant in Wellcome EPMC publications but not found to link to any grants in the grants data: ")
print(set(epmc_df['pmid']).difference(set(grant_ref['pmid'])))
```